openapi/openapi.go

```diff
@@ -0,0 +1,242 @@
+package openapi
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+
+	"github.com/pkg/errors"
+	yaml "gopkg.in/yaml.v2"
+)
+
+type decoder struct {
+	isYAML bool
+	src    *bytes.Buffer
+}
+
+func NewDecoder(src io.Reader) Decoder {
+	var buf bytes.Buffer
+	_, err := io.Copy(&buf, src)
+	if err != nil {
+		fmt.Printf("io.Copy err = %s\n", err)
+	}
+
+	var isYAML bool
+	switch src := src.(type) {
+	case *os.File:
+		// if it's a file, we can guess the payload formatting
+		// from the name of the file
+		switch filepath.Ext(src.Name()) {
+		case ".yaml", ".yml":
+			isYAML = true
+		}
+	default:
+		// Otherwise, we sniff the content.
+		b := bytes.TrimSpace(buf.Bytes())
+		if len(b) > 0 && b[0] != '{' { // if we don't have a JSON map, assume YAML
+			isYAML = true
+		}
+	}
+
+	return &decoder{
+		src:    &buf,
+		isYAML: isYAML,
+	}
+}
+
+func (d *decoder) Decode(v interface{}) error {
+	if d.src.Len() == 0 {
+		return errors.New(`empty source`)
+	}
+
+	if d.isYAML {
+		return yaml.Unmarshal(d.src.Bytes(), v)
+	}
+	return json.Unmarshal(d.src.Bytes(), v)
+}
+
+func fetchRemoteContent(u string) (io.Reader, error) {
+	res, err := http.Get(u)
+	if err != nil {
+		return nil, errors.Wrap(err, `failed to get remote content`)
+	}
+
+	if res.StatusCode != http.StatusOK {
+		return nil, errors.Errorf(`remote content responded with status %d`, res.StatusCode)
+	}
+
+	defer res.Body.Close()
+
+	var buf bytes.Buffer
+	if _, err := io.Copy(&buf, res.Body); err != nil {
+		return nil, errors.Wrap(err, `failed to read remote content`)
+	}
+
+	return &buf, nil
+}
+
+func Load(src io.Reader) (*Spec, error) {
+	var spec Spec
+
+	dec := NewDecoder(src)
+	if err := dec.Decode(&spec); err != nil {
+		return nil, errors.Wrap(err, `failed to decode content`)
+	}
+
+	// no paths or defs declared? check if this is a plain map[name]*Schema (definitions)
+	if len(spec.Paths) == 0 && len(spec.Definitions) == 0 {
+		var defs map[string]*Schema
+		if err := dec.Decode(&defs); err == nil {
+			if _, nok := defs["type"]; !nok {
+				spec.Definitions = defs
+			}
+		}
+	}
+
+	// _still_ no defs? try to see if this is a single item
+	// check if its just an *Item
+	if len(spec.Paths) == 0 && len(spec.Definitions) == 0 {
+		var item Schema
+		if err := dec.Decode(&item); err == nil {
+			spec.Definitions = map[string]*Schema{
+				"TODO": &item,
```

Comment:
This looks like it was originally used for resolving external references? Since I wrote external references to not use `openapi.LoadXXXX` functions, perhaps it's time this is deleted.

Reply:
☠️